class.two-factor-force.php

```diff
@@ -0,0 +1,488 @@
+						<?php self::global_force_2fa_by_role_field(); ?>
+					</td>
+				</tr>
+			</tbody>
+		</table>
+		<?php
+	}
+
+	/**
+	 * HTML output for global force 2fa field.
+	 */
+	public static function global_force_2fa_field() {
+		$is_universally_forced = self::get_universally_forced_option();
+
+		?>
+		<label>
+			<input type='checkbox' name="<?php echo esc_attr( self::FORCED_SITE_META_KEY ); ?>" value="1" <?php checked( $is_universally_forced ); ?> />
+			<?php esc_html_e( 'Force two-factor for all users', 'two-factor' ); ?>
+		</label>
+		<?php
+	}
+
+	/**
+	 * HTML output for per-role force 2fa fields.
+	 */
+	public static function global_force_2fa_by_role_field() {
+		$forced_roles          = self::get_forced_user_roles();
+		$is_universally_forced = self::get_universally_forced_option();
+
+		foreach ( get_editable_roles() as $slug => $role ) :
+			?>
+			<label>
+				<input type='checkbox' name="<?php echo esc_attr( sprintf( '%s[%s]', self::FORCED_ROLES_META_KEY, $slug ) ); ?>" value="1" <?php checked( in_array( $slug, $forced_roles, true ) ); ?> <?php echo ( $is_universally_forced ) ? 'readonly' : ''; ?> />
+				<?php echo esc_html( $role['name'] ); ?>
+			</label>
+			<br/>
+			<?php
+		endforeach;
+	}
+
+	/**
+	 * Save the force two_factor options against a network.
+	 *
+	 * @since 0.1-dev
+	 */
+	public static function save_network_force_two_factor_update() {
+		if ( ! isset( $_POST['_nonce_force_two_factor_options'] ) ) {
+			return;
+		}
+
+		check_admin_referer( 'force_two_factor_options', '_nonce_force_two_factor_options' );
+
+		// Validate and save universally forced key.
+		if ( isset( $_POST[ self::FORCED_SITE_META_KEY ] ) && $_POST[ self::FORCED_SITE_META_KEY ] ) {
+			update_site_option( self::FORCED_SITE_META_KEY, 1 );
+		} else {
+			update_site_option( self::FORCED_SITE_META_KEY, 0 );
+		}
+
+		// Validate and save per-role settings.
+		if (
+			! isset( $_POST[ self::FORCED_ROLES_META_KEY ] ) ||
+			! is_array( $_POST[ self::FORCED_ROLES_META_KEY ] )
+		) {
+			return;
+		}
+
+		// Whitelist roles against valid WordPress role slugs.
+		$roles = self::validate_forced_roles( $_POST[ self::FORCED_ROLES_META_KEY ] );
+
+		update_site_option( self::FORCED_ROLES_META_KEY, $roles );
+	}
+
+	/**
+	 * Validate and whitelist role values against valid editable_roles.
+	 *
+	 * @param array $unsafe_roles Values to validate and save.
+	 * @return array Whitelisted and safe values.
+	 */
+	public static function validate_forced_roles( $unsafe_roles ) {
+		$safe_roles = array_filter( wp_unslash( $unsafe_roles ), function( $is_role_saved, $role_slug ) {
+			return $is_role_saved && in_array( $role_slug, array_keys( get_editable_roles() ), true );
+		}, ARRAY_FILTER_USE_BOTH );
+
+		return array_keys( $safe_roles );
+	}
+
+	/**
+	 * Register settings for a single site.
+	 */
+	public static function register_single_site_force_2fa_options() {
+		// Add a new setting group for forcing 2fa in General Options.
+		add_settings_section(
```

Comment:
Should we move the plugin settings into own file since we could then re-use them for other parameters?

Reply:
Good plan 👍